lua/autorun/client/cl_devtools.lua

```diff
@@ -1,144 +1,171 @@
+            end
 
-        times[k] = newtime
-    end
+            local wasFoundDuringLookup = false
+            for structName, func in pairs(SWAMP_DEV.structEnvironments) do
+                local structLookupPath = "lua/" .. structName
 
-    for _, v in ipairs(folders) do
-        if (v == "server") then continue end
+                if filePath:find(structLookupPath) then
+                    local className = fileName:sub(1,-5) -- remove the .lua extension.
+                    loadFile(filePath, func(className) or func(subDir:gsub(".*/", "")))
+                    wasFoundDuringLookup = true
+                end
+            end
 
-        if (not times[v]) then
-            times[v] = {}
+            if not wasFoundDuringLookup then 
+                loadFile(filePath)
+            end
+            fileTimes[filePath] = fTime
         end
 
-        -- defining struct
-        if SWAMP_DEV.structTypes[v] then
-            SWAMP_DEV.itemStruct = v
-        end
-
-        times[v] = recurseRefresh(root, times[v], curPath .. v .. "/")
-
-        if (curPath == "") then
-            SWAMP_DEV.itemStruct = nil
+        for _,folder in ipairs(folders) do
+            refreshFolder(subDir .. "/" .. folder)
         end
     end
-
-    return times
 end
 
 -- toggle dev editing. Root is addons/contrib
 concommand.Add("dev", function()
-    if (LocalPlayer():GetRank() <= 0) then return end
-    SWAMP_DEV.isDev = not SWAMP_DEV.isDev
+    if not SWAMP_DEV.allowed[LocalPlayer():SteamID()] then return end
 
-    SWAMP_DEV.fileTimes = {
-        ["lua"] = {},
-        ["gamemodes"] = {}
-    }
+    SWAMP_DEV.enabled = not SWAMP_DEV.enabled
 
-    if (SWAMP_DEV.isDev) then
+    if SWAMP_DEV.enabled then
         print("Dev mode enabled")
 
         timer.Create("SWAMP_DEV.Refresh", SWAMP_DEV.refreshDelay, 0, function()
             -- avoid checking unnecessary folders
-            recurseRefresh("addons/contrib/lua/", SWAMP_DEV.fileTimes["lua"], "")
-            recurseRefresh("addons/contrib/gamemodes/", SWAMP_DEV.fileTimes["gamemodes"], "")
+            refreshFolder(ROOT_DIRECTORY .. "lua")
+            refreshFolder(ROOT_DIRECTORY .. "gamemodes")
+
         end)
     else
         print("Dev mode disabled")
         timer.Remove("SWAMP_DEV.Refresh")
     end
 end, nil, "Toggle dev mode, editing in addons/contrib", FCVAR_UNREGISTERED)
 
--- force refresh file
+-- force refresh a singular file 
 concommand.Add("dev_refresh", function(_, _, args)
-    if not SWAMP_DEV.isDev or not args[1] then return end
-    print("Attempting to force refresh file " .. args[1])
-
-    for s in string.gmatch(args[1], "[^/\\]+") do
-        if SWAMP_DEV.structTypes[s] then
-            SWAMP_DEV.itemStruct = s
+    local filePath = args[1]
+    if not SWAMP_DEV.enabled or not filePath then return end
+    print("Attempting to force refresh file " ..filePath)
+    
+    local structFound 
+    local ran
+    for fileType in filePath:gmatch("[^/\\]+") do
```

Comment:
Is there logic here that can be shared with the other refresh function? Maybe just have a flat list of all files, then whenever one is updated, do like RunConsoleCommand("dev_refresh", filename)

Reply:
Not really as this traverses the path, whereas, in the the automatic refresh it's loading each folder sepeartely.